### PR TITLE
Create Signed wheels with sigstore

### PR DESCRIPTION
**Pull request by andife**
### Description
Signing of software artifacts has become now a standard in secure software supply chains.
More details about sigstore 

- https://openssf.org/community/sigstore/, 
- https://docs.sigstore.dev/
- https://www.python.org/download/sigstore/

It is now also officially recommended in the documentation 
https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/
and the availability of github actions makes it easy to integrate.

Another point that still looks good in my view with "Configuring trusted publishing" is the elimination of the password and the direct configuration of the release in pypi. [That one can do at a different PR]

Other projects using sigstore: https://landscape.openssf.org/sigstore

### Cons?
* Additional files are created and published to github releases (example: https://github.com/andife/onnx/releases/tag/sigstoretest) => Space? (about 15MB per wheel, we have one for each python version?)

### Remark
* Only of one of the 4 release pipeline was adapted here

### CODE REVIEW

File: .github/workflows/release_linux_x86_64.yml [hunk 1]
```diff
@@ -19,7 +19,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
+        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
```
Comment (justinchuby) [inline .github/workflows/release_linux_x86_64.yml:22]: Trailing spaces

**Review by justinchuby — COMMENTED** [review]
LGTM! I would let others chime in as well

File: .github/workflows/release_linux_x86_64.yml [hunk 2]
```diff
@@ -98,3 +99,50 @@ jobs:
         export ORT_MAX_ML_OPSET_SUPPORTED_VERSION=3
         export ORT_MAX_ONNX_OPSET_SUPPORTED_VERSION=19
         pytest
+
+
```
Comment (justinchuby) [inline .github/workflows/release_linux_x86_64.yml:103]:
```suggestion
```

File: .github/workflows/release_linux_x86_64.yml [hunk 3]
```diff
@@ -98,3 +99,50 @@ jobs:
         export ORT_MAX_ML_OPSET_SUPPORTED_VERSION=3
         export ORT_MAX_ONNX_OPSET_SUPPORTED_VERSION=19
         pytest
+
+  
+  
+  github-release:
```
Comment (justinchuby) [inline .github/workflows/release_linux_x86_64.yml:105]: Would it make sense to upload signed packages in the `Upload wheel to PyPI weekly` step?

File: .github/workflows/release_linux_x86_64.yml [hunk 4]
```diff
@@ -98,3 +99,50 @@ jobs:
...
+  
+    runs-on: ubuntu-latest
+
+    permissions:
+      contents: write  # IMPORTANT: mandatory for making GitHub Releases
+      id-token: write  # IMPORTANT: mandatory for sigstore
+
+    steps:
+    - name: Download all the dists
+      uses: actions/download-artifact@v3
```
Comment (justinchuby) [inline .github/workflows/release_linux_x86_64.yml:120]: Should we sign the artifacts, or would it make sense to sign them before `Store distribution packages`?

File: .github/workflows/release_linux_x86_64.yml [hunk 5]
```diff
@@ -98,3 +99,50 @@ jobs:
...
+      with:
+        name: wheels
+        path: dist/
+    - name: Sign the dists with Sigstore
+      uses: sigstore/gh-action-sigstore-python@v1.2.3
+      with:
+        inputs: >-
+          ./dist/*.tar.gz
+          ./dist/*.whl
+    - name: Create GitHub Release
```
Comment (justinchuby) [inline .github/workflows/release_linux_x86_64.yml:130]: We may want to create releases only for the weekly and official packages, not for every build.

File: .github/workflows/release_linux_x86_64.yml [hunk 6]
```diff
@@ -98,3 +99,50 @@ jobs:
...
+      env:
+        GITHUB_TOKEN: ${{ github.token }}
+      run: >-
+        gh release create
+        '${{ github.ref_name }}'
+        --repo '${{ github.repository }}'
+        --notes ""
+    - name: Upload artifact signatures to GitHub Release
```
Comment (justinchuby) [inline .github/workflows/release_linux_x86_64.yml:138]: I would make sure all upload steps are guarded with conditions so we don't mistakenly upload test artifacts

**Comment by justinchuby** [comment]
> Configuring trusted publishing

That would be very helpful too!

**Comment by andife** [comment]
Integration directly into pip is also planed: https://peps.python.org/pep-0740/

**Pull request closed without merging**